File: app/src/main/java/com/cooper/wheellog/utils/FileUtil.java
package com.cooper.wheellog.utils;

import android.content.ContentResolver;
import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.os.Build;
import android.os.Environment;
import android.provider.MediaStore;

import androidx.annotation.RequiresApi;

import com.cooper.wheellog.views.TripModel;
import com.google.common.io.ByteStreams;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.Hashtable;
import java.util.Locale;
import java.util.Objects;

import timber.log.Timber;

public class FileUtil {
    private final Context context;
    private File file;
    private Uri uri;
    private final Hashtable<String, CachedFile> AndroidQCache = new Hashtable<>();
    private boolean ignoreTimber = false;
    private OutputStream stream;

    static class CachedFile {
        public File file;
        public Uri uri;
    }

    public FileUtil(Context context) {
        this.context = context;
    }

    public String fileName = "";

    public void setIgnoreTimber(boolean value) {
        ignoreTimber = value;
    }

    public File getFile() {
        return file;
    }

    public String getAbsolutePath() {
        if (isNull()) {
            return null;
        }

        return file.getAbsolutePath();
    }

    public boolean isNull() {
        return file == null || file.toString().equals("null") || stream == null;
    }

    public boolean prepareFile(String fileName) {
        return prepareFile(fileName, "");
    }

    public boolean prepareFile(String fileName, String folder) {
        this.fileName = fileName;
        uri = null;
        file = null;
        // Android 10+
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            // Get uri from cashed dictionary
            // It need to not create duplicate files
            if (AndroidQCache.containsKey(fileName)) {
                CachedFile cache = AndroidQCache.get(fileName);
                uri = cache.uri;
                file = cache.file;
                return true;
            }
            try {
                ContentValues contentValues = new ContentValues();
                contentValues.put(MediaStore.Downloads.DISPLAY_NAME, fileName);
                contentValues.put(MediaStore.Downloads.TITLE, fileName);
                contentValues.put(MediaStore.Downloads.MIME_TYPE, getMimeType(fileName));
                String path = Environment.DIRECTORY_DOWNLOADS + File.separator + Constants.LOG_FOLDER_NAME;
                if (folder != null && !folder.equals("")) {
                    path += File.separator + folder.replace(':', '_');
                }
                contentValues.put(MediaStore.Downloads.RELATIVE_PATH, path);
                Uri contentUri = MediaStore.Downloads.getContentUri(MediaStore.VOLUME_EXTERNAL_PRIMARY);

                uri = Objects.requireNonNull(getContentResolver()).insert(contentUri, contentValues);
                file = new File(getPathFromUri(uri));
            } finally {
                CachedFile cache = new CachedFile();
                cache.uri = uri;
                cache.file = file;
                AndroidQCache.put(fileName, cache);
            }
        } else {
            // api 28 or less
            // Get the directory for the user's public pictures directory.
            String path = Constants.LOG_FOLDER_NAME;
            if (folder != null && !folder.equals("")) {
                path += File.separator + folder.replace(':', '_');;
            }
            File dir = new File(Environment.getExternalStoragePublicDirectory(
                    Environment.DIRECTORY_DOWNLOADS), path);

            if (!dir.mkdirs() && !ignoreTimber)
                Timber.i("Directory not created");

            file = new File(dir, fileName);
        }
        prepareStream();
        return !isNull();
    }

    public void prepareStream() {
        try {
            close();
            if (uri != null) {
                stream = context.getContentResolver().openOutputStream(Objects.requireNonNull(uri), "wa");
            } else if (file != null) {
                stream = new FileOutputStream(file, true);
            }
        } catch (FileNotFoundException e) {
            if (!ignoreTimber) {
                Timber.e("File not found.");
            }
            //e.printStackTrace();
        }
    }

    public InputStream getInputStream() {
        try {
            if (uri != null) {
                return context.getContentResolver().openInputStream(Objects.requireNonNull(uri));
            } else if (file != null) {
                return new FileInputStream(file);
            }
        } catch (FileNotFoundException e) {
            if (!ignoreTimber) {
                Timber.e("File not found.");
            }
            e.printStackTrace();
        }
        return null;
    }

    public byte[] readBytes() throws IOException {
        InputStream inputStream = null;
        if (uri != null) {
            inputStream = context.getContentResolver().openInputStream(uri);
        } else if (file != null) {
            inputStream = new FileInputStream(file);
        }
        assert inputStream != null;
        return ByteStreams.toByteArray(inputStream);
    }

    public static byte[] readBytes(String filePath) throws IOException {
        InputStream inputStream = new FileInputStream(filePath);
        return ByteStreams.toByteArray(inputStream);
    }

    public static byte[] readBytes(Context context, Uri uri) throws IOException {
        InputStream inputStream = context.getContentResolver().openInputStream(uri);
        return ByteStreams.toByteArray(inputStream);
    }

    static String sizeTokb(Long size) {
        return String.format(Locale.US, "%.2f Kb", size / 1024f);
    }

    public static FileUtil getLastLog(Context context) {
        String fileStartsWith = new SimpleDateFormat("yyyy_MM_dd", Locale.US).format(new Date());

        // Android 9 or less
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q)
        {
            File dir = new File(Environment.getExternalStoragePublicDirectory(
                    Environment.DIRECTORY_DOWNLOADS), Constants.LOG_FOLDER_NAME);
            File[] filesArray = dir.listFiles();
            if (filesArray == null) {
                return null;
            }
            for (File wheelDir: filesArray) {
                if (wheelDir.isDirectory()) {
                    File[] wheelFiles = wheelDir.listFiles();
                    if (wheelFiles == null) {
                        continue;
                    }
                    for (File f: wheelFiles) {
                        int indexExt = f.getAbsolutePath().lastIndexOf(".");
                        if (f.isDirectory() || indexExt < 1) {
                            continue;
                        }
                        String extension = f.getAbsolutePath().substring(indexExt);
                        if (extension.equals(".csv") && f.getName().startsWith(fileStartsWith)) {
                            FileUtil result = new FileUtil(context);
                            result.file = f;
                            result.fileName = f.getName();
                            return result;
                        }
                    }
                }
            }
            return null;
        }
        // Android 10+
        Uri uri = MediaStore.Downloads.getContentUri(MediaStore.VOLUME_EXTERNAL_PRIMARY);
        String[] projection = {
                MediaStore.Downloads.MIME_TYPE,
                MediaStore.Downloads.DISPLAY_NAME,
                MediaStore.Downloads.TITLE,
                MediaStore.Downloads.SIZE,
                MediaStore.Downloads._ID
        };
        String where = String.format("%s = 'text/comma-separated-values'", MediaStore.Downloads.MIME_TYPE);
        Cursor cursor = context.getContentResolver().query(uri,
                projection,
                where + " AND " + MediaStore.Downloads.DISPLAY_NAME + " LIKE ?",
                new String[] { fileStartsWith + "%" },
                MediaStore.Downloads.DATE_MODIFIED + " DESC");
        if (cursor != null && cursor.moveToFirst()) {
            String title = cursor.getString(cursor.getColumnIndex(MediaStore.Downloads.DISPLAY_NAME));
            String mediaId = cursor.getString(cursor.getColumnIndex(MediaStore.Downloads._ID));
            cursor.close();
            FileUtil result = new FileUtil(context);
            Uri downloads = MediaStore.Downloads.getContentUri(MediaStore.VOLUME_EXTERNAL);
            result.uri = Uri.withAppendedPath(downloads, mediaId);
            result.file = new File(result.getPathFromUri(result.uri));
            result.fileName = title;
            return result;
        }
        return null;
    }

    public static ArrayList<TripModel> fillTrips(Context context) {
        ArrayList<TripModel> tripModels = new ArrayList<>();
        // Android 9 or less
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q)
        {
            File dir = new File(Environment.getExternalStoragePublicDirectory(
                    Environment.DIRECTORY_DOWNLOADS), Constants.LOG_FOLDER_NAME);
            File[] filesArray = dir.listFiles();
            if (filesArray == null) {
                return tripModels;
            }
            for (File wheelDir: filesArray) {
                if (wheelDir.isDirectory()) {
                    File[] wheelFiles = wheelDir.listFiles();
                    if (wheelFiles == null) {
                        continue;
                    }
                    for (File f: wheelFiles) {
                        int indexExt = f.getAbsolutePath().lastIndexOf(".");
                        if (f.isDirectory() || indexExt < 1) {
                            continue;
                        }
                        String extension = f.getAbsolutePath().substring(indexExt);
                        if (extension.equals(".csv") && !f.getName().startsWith("RAW")) {
                            tripModels.add(new TripModel(f.getName(), sizeTokb(f.length()), f.getAbsolutePath()));
                        }
                    }
                }
            }
            return tripModels;
        }
        // Android 10+
        Uri uri = MediaStore.Downloads.getContentUri(MediaStore.VOLUME_EXTERNAL_PRIMARY);
        String[] projection = {
                MediaStore.Downloads.MIME_TYPE,
                MediaStore.Downloads.DISPLAY_NAME,
                MediaStore.Downloads.TITLE,
                MediaStore.Downloads.SIZE,
                MediaStore.Downloads._ID
        };
        String where = String.format("%s = 'text/comma-separated-values'", MediaStore.Downloads.MIME_TYPE);
        Cursor cursor = context.getContentResolver().query(uri,
                projection,
                where + " AND " + MediaStore.Downloads.DISPLAY_NAME + " NOT LIKE ?",
                new String[] { "RAW_%" },
                MediaStore.Downloads.DATE_MODIFIED + " DESC");
        if (cursor != null && cursor.moveToFirst()) {
            do {
                String title = cursor.getString(cursor.getColumnIndex(MediaStore.Downloads.DISPLAY_NAME));
                String description = sizeTokb(cursor.getLong(cursor.getColumnIndex(MediaStore.Downloads.SIZE)));
                String mediaId = cursor.getString(cursor.getColumnIndex(MediaStore.Downloads._ID));
                tripModels.add(new TripModel(title, description, mediaId));
            } while (cursor.moveToNext());
            cursor.close();
        }
        return tripModels;
    }

    public void close() {
        if (stream == null) {
            return;
        }
        try {
            stream.close();
            stream = null;
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void writeLine(String line) {
        if (isNull()) {
            if (!ignoreTimber) {
                Timber.e("Write failed. File is null");
            }
            return;
        }
        if (stream == null) {
            if (!ignoreTimber) {
                Timber.e("Write failed. Stream is null. Forgot to call prepareStream()?");
            }
            return;
        }

        try {
            stream.write((line + "\r\n").getBytes());
            stream.flush();
        } catch (IOException e) {
            if (!ignoreTimber) {
                Timber.e("IOException");
            }
            e.printStackTrace();
        }
    }

    @Nullable
    private ContentResolver getContentResolver() {
        if (context != null) {
            return context.getContentResolver();
        } else {
            return null;
        }
    }

    @NotNull
    private String getMimeType(@NotNull String fileName) {
        if (fileName.endsWith(".csv")) {
            return "text/csv";
        } else if (fileName.endsWith(".html") || fileName.endsWith(".htm")) {
            return "text/html";
        } else {
            return "*/*";
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.KITKAT)
    protected String getPathFromUri(@Nullable Uri uri) {
        if (uri == null) {
            return "";
        }
        if ("content".equalsIgnoreCase(uri.getScheme())) {
            return getDataColumn();
        }
        if ("file".equalsIgnoreCase(uri.getScheme())) {
            return uri.getPath();
        }

        return "";
    }

    private String getDataColumn() {
        Cursor cursor = null;
        final String column = "_data";
        final String[] projection = {
                column
        };

        try {
            cursor = context.getContentResolver().query(uri, projection, null, null,
                    null);
            if (cursor != null && cursor.moveToFirst()) {
                final int index = cursor.getColumnIndexOrThrow(column);
                return cursor.getString(index);
            }
        } finally {
            if (cursor != null)
                cursor.close();
        }
        return "";
    }

    @Override
    protected void finalize() throws Throwable {
        close();
        super.finalize();
    }
}

File: app/src/main/java/com/cooper/wheellog/utils/FileUtil.kt
package com.cooper.wheellog.utils

import android.content.ContentResolver
import android.content.ContentValues
import android.content.Context
import android.content.Intent
import android.database.Cursor
import android.net.Uri
import android.os.Build
import android.os.Environment
import android.provider.MediaStore
import android.widget.Toast
import androidx.core.content.FileProvider
import androidx.preference.PreferenceManager
import com.cooper.wheellog.data.Consts
import com.cooper.wheellog.views.TripModel
import com.google.common.io.ByteStreams
import org.json.JSONArray
import org.json.JSONObject
import timber.log.Timber
import java.io.*
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Hashtable
import java.util.Locale
import java.util.zip.ZipEntry
import java.util.zip.ZipInputStream
import java.util.zip.ZipOutputStream

class FileUtil(val context: Context) {
    var file: File? = null
        private set
    private var uri: Uri? = null
    private val androidQCache = Hashtable<String, CachedFile>()
    private var ignoreTimber = false
    private var stream: OutputStream? = null

    internal class CachedFile {
        var file: File? = null
        var uri: Uri? = null
    }

    var fileName = ""
    fun setIgnoreTimber(value: Boolean) {
        ignoreTimber = value
    }

    val absolutePath: String?
        get() = if (isNull) {
            null
        } else file!!.absolutePath
    val isNull: Boolean
        get() = file == null || file.toString() == "null" || stream == null

    @JvmOverloads
    fun prepareFile(fileName: String, folder: String? = ""): Boolean {
        this.fileName = fileName
        uri = null
        file = null
        // Android 10+
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            // Get uri from cashed dictionary
            // It need to not create duplicate files
            if (androidQCache.containsKey(fileName)) {
                val cache = androidQCache[fileName]
                uri = cache!!.uri
                file = cache.file
                return true
            }
            try {
                val contentValues = ContentValues()
                contentValues.put(MediaStore.Downloads.DISPLAY_NAME, fileName)
                contentValues.put(MediaStore.Downloads.TITLE, fileName)
                contentValues.put(MediaStore.Downloads.MIME_TYPE, getMimeType(fileName))
                var path =
                    Environment.DIRECTORY_DOWNLOADS + File.separator + Constants.LOG_FOLDER_NAME
                if (folder != null && folder != "") {
                    path += File.separator + folder.replace(':', '_')
                }
                contentValues.put(MediaStore.Downloads.RELATIVE_PATH, path)
                val contentUri =
                    MediaStore.Downloads.getContentUri(MediaStore.VOLUME_EXTERNAL_PRIMARY)
                uri = contentResolver?.insert(contentUri, contentValues)
                file = getPathFromUri(uri)?.let { File(it) }
            } finally {
                val cache = CachedFile()
                cache.uri = uri
                cache.file = file
                androidQCache[fileName] = cache
            }
        } else {
            // api 28 or less
            // Get the directory for the user's public pictures directory.
            var path = Constants.LOG_FOLDER_NAME
            if (folder != null && folder != "") {
                path += File.separator + folder.replace(':', '_')
            }
            val dir = File(
                Environment.getExternalStoragePublicDirectory(
                    Environment.DIRECTORY_DOWNLOADS
                ), path
            )
            if (!dir.mkdirs() && !ignoreTimber) Timber.i("Directory not created")
            file = File(dir, fileName)
        }
        prepareStream()
        return !isNull
    }

    fun prepareStream() {
        try {
            close()
            if (uri != null) {
                stream = context.contentResolver.openOutputStream(uri!!, "wa")
            } else if (file != null) {
                stream = FileOutputStream(file, true)
            }
        } catch (e: FileNotFoundException) {
            if (!ignoreTimber) {
                Timber.e("File not found.")
            }
            //e.printStackTrace();
        }
    }

    val inputStream: InputStream?
        get() {
            try {
                if (uri != null) {
                    return context.contentResolver.openInputStream(uri!!)
                } else if (file != null) {
                    return FileInputStream(file)
                }
            } catch (e: FileNotFoundException) {
                if (!ignoreTimber) {
                    Timber.e("File not found.")
                }
                e.printStackTrace()
            }
            return null
        }

    @Throws(IOException::class)
    fun readBytes(): ByteArray? {
        var inputStream: InputStream? = null
        if (uri != null) {
            inputStream = context.contentResolver.openInputStream(uri!!)
        } else if (file != null) {
            inputStream = FileInputStream(file)
        }
        return inputStream?.let { ByteStreams.toByteArray(it) }
    }

    fun close() {
        if (stream == null) {
            return
        }
        try {
            stream!!.close()
            stream = null
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun writeLine(line: String) {
        if (isNull) {
            if (!ignoreTimber) {
                Timber.wtf("Write failed. File is null")
            }
            return
        }
        if (stream == null) {
            if (!ignoreTimber) {
                Timber.wtf("Write failed. Stream is null. Forgot to call prepareStream()?")
            }
            return
        }
        try {
            stream!!.write((line + "\r\n").toByteArray())
            stream!!.flush()
        } catch (e: IOException) {
            if (!ignoreTimber) {
                Timber.wtf("IOException %s", e.message)
            }
            e.printStackTrace()
        }
    }

    private val contentResolver: ContentResolver?
        get() = context.contentResolver

    private fun getMimeType(fileName: String): String {
        return if (fileName.endsWith(".csv")) {
            "text/csv"
        } else if (fileName.endsWith(".html") || fileName.endsWith(".htm")) {
            "text/html"
        } else {
            "*/*"
        }
    }

    fun getPathFromUri(uri: Uri?): String? {
        if (uri == null) {
            return ""
        }
        if ("content".equals(uri.scheme, ignoreCase = true)) {
            return dataColumn
        }
        return if ("file".equals(uri.scheme, ignoreCase = true)) {
            uri.path
        } else ""
    }

    private val dataColumn: String
        get() {
            var cursor: Cursor? = null
            val column = "_data"
            val projection = arrayOf(
                column
            )
            try {
                cursor = context.contentResolver.query(
                    uri!!, projection, null, null,
                    null
                )
                if (cursor != null && cursor.moveToFirst()) {
                    val index = cursor.getColumnIndexOrThrow(column)
                    return cursor.getString(index)
                }
            } finally {
                cursor?.close()
            }
            return ""
        }

    @Throws(Throwable::class)
    protected fun finalize() {
        close()
    }

    companion object {
        private fun sizeToKb(size: Long): String {
            return String.format(Locale.US, "%.2f Kb", size / 1024f)
        }

        fun getLastLog(context: Context): FileUtil? {
            val fileStartsWith = SimpleDateFormat("yyyy_MM_dd", Locale.US).format(Date())

            // Android 9 or less
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
                val dir = File(
                    Environment.getExternalStoragePublicDirectory(
                        Environment.DIRECTORY_DOWNLOADS
                    ), Constants.LOG_FOLDER_NAME
                )
                val filesArray = dir.listFiles() ?: return null
                for (wheelDir in filesArray) {
                    if (wheelDir.isDirectory) {
                        val wheelFiles = wheelDir.listFiles() ?: continue
                        for (f in wheelFiles) {
                            val indexExt = f.absolutePath.lastIndexOf(".")
                            if (f.isDirectory || indexExt < 1) {
                                continue
                            }
                            val extension = f.absolutePath.substring(indexExt)
                            if (extension == ".csv" && f.name.startsWith(fileStartsWith)) {
                                val result = FileUtil(context)
                                result.file = f
                                result.fileName = f.name
                                return result
                            }
                        }
                    }
                }
                return null
            }
            // Android 10+
            val uri = MediaStore.Downloads.getContentUri(MediaStore.VOLUME_EXTERNAL_PRIMARY)
            val projection = arrayOf(
                MediaStore.Downloads.MIME_TYPE,
                MediaStore.Downloads.DISPLAY_NAME,
                MediaStore.Downloads.TITLE,
                MediaStore.Downloads.SIZE,
                MediaStore.Downloads._ID
            )
            val where =
                String.format("%s = 'text/comma-separated-values'", MediaStore.Downloads.MIME_TYPE)
            val cursor = context.contentResolver.query(
                uri,
                projection,
                where + " AND " + MediaStore.Downloads.DISPLAY_NAME + " LIKE ?", arrayOf(
                    "$fileStartsWith%"
                ),
                MediaStore.Downloads.DATE_MODIFIED + " DESC"
            )
            if (cursor != null && cursor.moveToFirst()) {
                val title =
                    cursor.getString(0.coerceAtLeast(cursor.getColumnIndex(MediaStore.Downloads.DISPLAY_NAME)))
                val mediaId =
                    cursor.getString(0.coerceAtLeast(cursor.getColumnIndex(MediaStore.Downloads._ID)))
                cursor.close()
                val result = FileUtil(context)
                val downloads = MediaStore.Downloads.getContentUri(MediaStore.VOLUME_EXTERNAL)
                result.uri = Uri.withAppendedPath(downloads, mediaId)
                result.file = result.getPathFromUri(result.uri)?.let { File(it) }
                result.fileName = title
                return result
            }
            return null
        }

        fun fillTrips(context: Context): ArrayList<TripModel> {
            val tripModels = ArrayList<TripModel>()
            // Android 9 or less
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
                val dir = File(
                    Environment.getExternalStoragePublicDirectory(
                        Environment.DIRECTORY_DOWNLOADS
                    ), Constants.LOG_FOLDER_NAME
                )
                var filesArray = dir.listFiles() ?: arrayOf<File>()

                val dirOld = File(
                    Environment.getExternalStoragePublicDirectory(
                        Environment.DIRECTORY_DOWNLOADS
                    ), Constants.LOG_FOLDER_OLD_NAME
                )
                filesArray += dirOld.listFiles() ?: arrayOf<File>()
                if (filesArray.isEmpty()) {
                    return tripModels
                }
                for (wheelDir in filesArray) {
                    if (wheelDir.isDirectory) {
                        val wheelFiles = wheelDir.listFiles() ?: continue
                        for (f in wheelFiles) {
                            val indexExt = f.absolutePath.lastIndexOf(".")
                            if (f.isDirectory || indexExt < 1) {
                                continue
                            }
                            val extension = f.absolutePath.substring(indexExt)
                            if (extension == ".csv" && !f.name.startsWith("RAW")) {
                                tripModels.add(
                                    TripModel(
                                        f.name,
                                        sizeToKb(f.length()),
                                        f.absolutePath
                                    )
                                )
                            }
                        }
                    }
                }
                return tripModels
            }
            // Android 10+
            val uri = MediaStore.Downloads.getContentUri(MediaStore.VOLUME_EXTERNAL_PRIMARY)
            val projection = arrayOf(
                MediaStore.Downloads.MIME_TYPE,
                MediaStore.Downloads.DISPLAY_NAME,
                MediaStore.Downloads.TITLE,
                MediaStore.Downloads.SIZE,
                MediaStore.Downloads._ID
            )
            val where =
                String.format("%s = 'text/comma-separated-values'", MediaStore.Downloads.MIME_TYPE)
            val cursor = context.contentResolver.query(
                uri,
                projection,
                where + " AND " + MediaStore.Downloads.DISPLAY_NAME + " NOT LIKE ?",
                arrayOf("RAW_%"),
                MediaStore.Downloads.DATE_MODIFIED + " DESC"
            )
            if (cursor != null && cursor.moveToFirst()) {
                do {
                    val title =
                        cursor.getString(0.coerceAtLeast(cursor.getColumnIndex(MediaStore.Downloads.DISPLAY_NAME)))
                    val description =
                        sizeToKb(cursor.getLong(0.coerceAtLeast(cursor.getColumnIndex(MediaStore.Downloads.SIZE))))
                    val mediaId =
                        cursor.getString(0.coerceAtLeast(cursor.getColumnIndex(MediaStore.Downloads._ID)))
                    tripModels.add(TripModel(title, description, mediaId))
                } while (cursor.moveToNext())
                cursor.close()
            }
            return tripModels
        }

        private const val BACKUP_FILE_NAME = "wheellog_backup.zip"
        private const val PREFS_FILE_NAME = "preferences.json"

        fun exportData(context: Context): File? {
            val dbPath = context.getDatabasePath(Consts.DATABASE_NAME).absolutePath
            val filesDir = context.filesDir.absolutePath
            val exportFile = File(context.getExternalFilesDir(Environment.DIRECTORY_DOCUMENTS), BACKUP_FILE_NAME)

            return try {
                ZipOutputStream(FileOutputStream(exportFile)).use { zipOut ->
                    addFileToZip(dbPath, "database/${Consts.DATABASE_NAME}", zipOut)
                    val files = File(filesDir).listFiles()
                    files?.forEach { file ->
                        addFileToZip(file.absolutePath, "files/${file.name}", zipOut)
                    }
                    // Экспортируем настройки
                    val prefsFile = File(context.cacheDir, PREFS_FILE_NAME)
                    savePreferencesToFile(context, prefsFile)
                    addFileToZip(prefsFile.absolutePath, "preferences/$PREFS_FILE_NAME", zipOut)

                }
                exportFile
            } catch (e: Exception) {
                e.printStackTrace()
                null
            }
        }

        /**
         * Импортирует данные из архива .zip
         */
        fun importData(context: Context, backupUri: Uri) {
            val dbPath = context.getDatabasePath(Consts.DATABASE_NAME).absolutePath
            val filesDir = context.filesDir.absolutePath

            try {
                val inputStream = context.contentResolver.openInputStream(backupUri) ?: return
                ZipInputStream(inputStream).use { zipIn ->
                    var entry = zipIn.nextEntry
                    while (entry != null) {
                        val outputFile = if (entry.name.startsWith("database/")) {
                            File(dbPath)
                        } else {
                            File(filesDir, entry.name.removePrefix("files/"))
                        }

                        FileOutputStream(outputFile).use { fos ->
                            zipIn.copyTo(fos)
                        }
                        zipIn.closeEntry()
                        entry = zipIn.nextEntry
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        /**
         * Добавляет файл в архив .zip
         */
        private fun addFileToZip(filePath: String, zipEntryName: String, zipOut: ZipOutputStream) {
            val file = File(filePath)
            if (!file.exists() || !file.isFile) return

            try {
                FileInputStream(file).use { fis ->
                    zipOut.putNextEntry(ZipEntry(zipEntryName))
                    fis.copyTo(zipOut)
                    zipOut.closeEntry()
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        /**
         * Сохранение всех SharedPreferences в JSON-файл
         */
        private fun savePreferencesToFile(context: Context, file: File) {
            val prefs = PreferenceManager.getDefaultSharedPreferences(context)
            val allPrefs = prefs.all
            val jsonObject = JSONObject()

            for ((key, value) in allPrefs) {
                when (value) {
                    is Boolean, is Int, is Float, is Double, is Long, is String -> jsonObject.put(key, value)
                    is Set<*> -> jsonObject.put(key, JSONArray(value))
                }
            }

            file.writeText(jsonObject.toString())
        }

        /**
         * Восстановление SharedPreferences из JSON-файла
         */
        private fun restorePreferencesFromFile(context: Context, file: File) {
            val prefs = PreferenceManager.getDefaultSharedPreferences(context)
            val editor = prefs.edit()

            val json = JSONObject(file.readText())
            json.keys().forEach { key ->
                when (val value = json.get(key)) {
                    is Boolean -> editor.putBoolean(key, value)
                    is Int -> editor.putInt(key, value)
                    is Float -> editor.putFloat(key, value)
                    is Double -> editor.putFloat(key, value.toFloat())
                    is Long -> editor.putLong(key, value)
                    is String -> editor.putString(key, value)
                    is JSONArray -> {
                        val set = mutableSetOf<String>()
                        for (i in 0 until value.length()) {
                            set.add(value.getString(i))
                        }
                        editor.putStringSet(key, set)
                    }
                }
            }
            editor.apply()
        }

        /**
         * Делится файлом через другие приложения
         */
        fun shareBackupFile(context: Context, file: File) {
            if (!file.exists()) {
                Timber.e("Файл не найден: ${file.absolutePath}")
                Toast.makeText(context, "Файл не найден", Toast.LENGTH_SHORT).show()
                return
            }

            try {
                val authority = "${context.packageName}.provider"
                val uri = FileProvider.getUriForFile(context, authority, file)

                val intent = Intent(Intent.ACTION_SEND).apply {
                    type = "application/zip"
                    putExtra(Intent.EXTRA_STREAM, uri)
                    addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                }

                context.startActivity(Intent.createChooser(intent, "Отправить резервную копию"))
            } catch (e: Exception) {
                Timber.e("Ошибка при передаче файла", e)
                Toast.makeText(context, "Не удалось отправить файл", Toast.LENGTH_SHORT).show()
            }
        }
    }
}
